File: Sources/LLMCore/Endpoint.swift
import Foundation

/// Base URL 문자열과 API 경로를 이중 슬래시 없이 합친다.
/// 사용자가 설정에 입력한 값이라 끝 슬래시/공백이 섞여 들어올 수 있다.
public func joinEndpoint(base: String, path: String) -> String {
    var b = base.trimmingCharacters(in: .whitespacesAndNewlines)
    while b.hasSuffix("/") { b.removeLast() }
    let p = path.hasPrefix("/") ? path : "/" + path
    return b + p
}

/// Base URL 문자열에서 scheme://host[:port] origin만 추출 (경로 제거). 실패 시 nil.
/// Ollama 네이티브 API(/api/*)는 /v1 경로 없이 origin에 붙는다.
public func endpointOrigin(_ base: String) -> String? {
    guard let url = URL(string: base.trimmingCharacters(in: .whitespacesAndNewlines)),
          let scheme = url.scheme, let host = url.host else { return nil }
    // url.host는 IPv6 리터럴의 대괄호를 벗겨 반환한다("::1") — 재부착하지 않으면
    // "http://::1:11434"처럼 포트와 뒤엉킨 잘못된 origin이 나온다.
    let bracketed = host.contains(":") ? "[\(host)]" : host
    let port = url.port.map { ":\($0)" } ?? ""
    return "\(scheme)://\(bracketed)\(port)"
}

File: windows/AnywhereLLM.Core/Endpoint.cs
namespace AnywhereLLM.Core;

/// Base-URL / path joining and origin extraction. The base URL is user-entered
/// (settings), so trailing slashes and whitespace can leak in. Ports the Swift
/// joinEndpoint / endpointOrigin free functions.
public static class Endpoint
{
    /// Join base URL and API path without a double slash.
    public static string Join(string @base, string path)
    {
        var b = @base.Trim();
        while (b.EndsWith("/", StringComparison.Ordinal)) b = b[..^1];
        var p = path.StartsWith("/", StringComparison.Ordinal) ? path : "/" + path;
        return b + p;
    }

    /// scheme://host[:port] origin only (path stripped). null on parse failure.
    /// Ollama's native API (/api/*) attaches to the origin with no /v1 path.
    // ponytail: GetLeftPart(Authority) drops a default port (e.g. :80/:443) that was
    // explicit in the input; the Swift version kept it. No behavior depends on it
    // (used only to reach /api/version and rebuild /api/chat), so this is fine.
    public static string? Origin(string @base)
    {
        if (!Uri.TryCreate(@base.Trim(), UriKind.Absolute, out var url)) return null;
        if (string.IsNullOrEmpty(url.Scheme) || string.IsNullOrEmpty(url.Host)) return null;
        return url.GetLeftPart(UriPartial.Authority);
    }
}
